bundles/org.openhab.binding.withings/src/main/java/org/openhab/binding/withings/internal/handler/WithingsBridgeHandler.java

```diff
@@ -0,0 +1,239 @@
+                    accessTokenService.importAccessToken(accessTokenResponse);
+                    return true;
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                            "Error on fetching refresh token! Please set an auth-code within the Withings API thing!");
+                }
+            } catch (OAuthException e) {
+                logger.error("Error on importing refresh token! Please configure a new auth-code. Message: {}",
+                        e.getMessage());
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Can't connect to Withings API. Please set an auth-code within the Withings API thing!");
+            } finally {
+                clearAuthCode();
+            }
+        }
+        return false;
+    }
+
+    private void refreshAccessToken(@Nullable String refreshToken) {
+        if (refreshToken == null) {
+            logger.warn("There is no valid refresh token! Please configure a new auth-code.");
+            return;
+        }
+
+        logger.debug("Refreshing the access token...");
+        try {
+
+            AuthHandler authHandler = new AuthHandler(accessTokenService, httpClient);
+            Optional<WithingsAccessTokenResponse> tokenResponse = authHandler.refreshAccessToken(configuration.clientId,
+                    configuration.clientSecret, refreshToken);
+
+            if (tokenResponse.isPresent()) {
+                AccessTokenResponse accessTokenResponse = tokenResponse.get().createAccessTokenResponse();
+                accessTokenService.importAccessToken(accessTokenResponse);
+            } else {
+                logger.error("Invalid access token response at refresh!");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Error on refreshing access token! Please set a new auth-code within the Withings API thing!");
+            }
+        } catch (OAuthException e) {
+            logger.warn("Error refreshing access token! Detail: {}", e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Can't connect to Withings API. Refreshing access token failed!");
+        }
+    }
+
+    private void clearAuthCode() {
+        final Configuration configuration = editConfiguration();
+        configuration.put(WithingsBindingConstants.CONFIG_AUTH_CODE, "");
+        updateConfiguration(configuration);
+
+        this.configuration = getConfigAs(WithingsBridgeConfiguration.class);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Running dispose()");
+
+        @Nullable
+        ScheduledFuture<?> job = tokenRefreshJob;
+        if (job != null && !job.isCancelled()) {
```

Comment:
No harm in canceling a task that is already cancelled.
```suggestion
        if (job != null) {
```

Reply:
Hm, the JavaDoc of the cancel method describes that it will fail when the job is already canceled. I would expect that "fail" means an exception or at least a confusing unnecessary log entry.

Reply:
The attempt will fail, but the method invocation will succeed.